File: sort/quick_sort/quick_sort.cpp
//
// Created by 韩大大 on 2022/8/14.
//

#include <iostream>

using namespace std;

const int N = 100010;

int q[N];

void quickSort(int q[], int l, int r) {
    if (l >= r) return;

    int i = l - 1, j = r + 1, x = q[l + r >> 1];
    while (i < j) {
        do {
            ++i;
        } while (q[i] < x);
        do {
            --j;
        } while (q[j] > x);

        if (i < j) {
            swap(q[i], q[j]);
        }
    }

    quickSort(q, l, j);
    quickSort(q, j + 1, r);
}

int main() {
    int n;
    scanf("%d", &n);

    for (int i = 0; i < n; i++) {
        scanf("%d", &q[i]);
    }

    quickSort(q, 0, n - 1);

    for (int i = 0; i < n; i++) {
        printf("%d ", q[i]);
    }
}

File: sort/quick_sort/quick_sort.go
/**
  @author: hs
  @date: 2023/10/22
  @note:

modification history
--------------------
**/
package main

import "fmt"

func main() {
	arr := []int{1, 5, 7, 4, 3, 3}
	quickSort(arr)
	fmt.Println("arr: ", arr)
}

func quickSort(arr []int) []int {
	return _quickSort(arr, 0, len(arr)-1)
}

func _quickSort(arr []int, left, right int) []int {
	if left < right {
		partitionIndex := partition(arr, left, right)
		_quickSort(arr, left, partitionIndex-1)
		_quickSort(arr, partitionIndex+1, right)
	}

	return arr
}

func partition(arr []int, left, right int) int {
	pivot := left
	index := left + 1

	for i := index; i <= right; i++ {
		if arr[i] < arr[pivot] {
			swap(arr, index, i)
			index++
		}
	}
	swap(arr, pivot, index-1)

	return index-1
}

func swap(arr []int, i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}